tao_compiler/mlir/disc/transforms/disc_transpose_simplifier.cc

```diff
@@ -0,0 +1,851 @@
+      input_has_tranpose_consumer = true;
+      continue;
+    }
+    if (candidate->getOperand(0) == op->getResult(0)) {
+      output_has_tranpose_consumer = true;
+      continue;
+    }
+
+    DenseMap<Operation*, SmallVector<int64_t>> intermedateOpsPermutationMap;
+    if (ctx.dominates(otherVal, candidate->getOperand(0), perm,
+                      intermedateOpsPermutationMap)) {
+      input_has_tranpose_consumer = true;
+      permMaps.push_back(intermedateOpsPermutationMap);
+    }
+    intermedateOpsPermutationMap.clear();
+    if (ctx.dominates(op->getResult(0), candidate->getOperand(0), perm,
+                      intermedateOpsPermutationMap)) {
+      output_has_tranpose_consumer = true;
+      permMaps.push_back(intermedateOpsPermutationMap);
+    }
+  }
+
+  if (!input_has_tranpose_consumer || !output_has_tranpose_consumer)
+    return success();
+  for (auto& permMap : permMaps)
+    if (failed(ctx.rewriteIntermediateOps(permMap))) return failure();
+
+  OpBuilder b(op);
+  SmallVector<Value> newOperands;
+  newOperands.push_back(insertTranspose(op, lhs, perm, b)->getResult(0));
+  newOperands.push_back(insertTranspose(op, rhs, perm, b)->getResult(0));
+  Operation* clonedOp = b.clone(*op);
+  clonedOp->setOperands(newOperands);
+  Type newType = getTransposeOutputType(op->getResult(0), perm, b);
+  clonedOp->getResult(0).setType(newType);
+  Value newResult =
+      insertTranspose(op, clonedOp->getResult(0), reversePerm, b)->getResult(0);
+  op->getResult(0).replaceAllUsesWith(newResult);
+
+  // convert:
+  //   x -> bcast -> tranpose
+  // to:
+  //   x -> transpose -> bcast
+  // in case having implicit bcast
+  if (implicit_bcast) {
+    Value newOperand = transposeLHS ? newOperands[1] : newOperands[0];
+    Value transposeOperand = newOperand.getDefiningOp()->getOperand(0);
+    TransposeSimpliferContext ctx(block);
+    DenseMap<Operation*, SmallVector<int64_t>> intermedateOpsPermutationMap;
+    bool status = ctx.dominates(otherVal, transposeOperand, perm,
+                                intermedateOpsPermutationMap);
+    (void)status;
+    assert(status);
+    if (failed(ctx.rewriteIntermediateOps(intermedateOpsPermutationMap)))
+      return failure();
+  }
+
+  changed = true;
+  return success();
+}
+
+LogicalResult reverseBinaryOpsIfBenefitial(Block* block, bool& changed) {
+  SmallVector<Operation*> ops;
+  for (Operation& op : *block) {
+    if (op.getDialect() != op.getContext()->getLoadedDialect("mhlo") &&
+        op.getDialect() != op.getContext()->getLoadedDialect("mhlo_disc")) {
+      continue;
+    }
+    if (!op.hasTrait<mlir::OpTrait::Elementwise>()) continue;
+    if (op.getNumOperands() != 2) continue;
+    ops.push_back(&op);
+  }
+
+  for (Operation* op : ops) {
+    if (failed(reverseIfOperandsAreConsistentTransposeOps(op, changed)))
+      return failure();
+    if (changed) return success();
+    if (failed(reverseIfOperandsAndResultsAreConsistent(op, changed)))
+      return failure();
```

Comment:
reverseIfOperandsAreConsistentTransposeOps & reverseIfOperandsAndResultsAreConsistent

Can these two patterns also be regarded as a general optimization, and put into runCanonicalizer()?
These patterns might be further moved into hlo_ops.cc.

Reply:
it may not be a good idea due to following reasons：

1， it's beneficial from the perspective of local view, while may not be beneficial from the perspective of global view.

2,   These patterns rely on relative expensive context analysis. We will have to pay for this always if we put these patterns to canonicalization pass.